File: generateFlowAccounts.py
#!/usr/bin/env python3
"""
Flow Wallet Generator for All Users

This script generates Flow wallets for all users in Supabase auth:
1. Fetches all users from Supabase auth using service role
2. Skips users who already have wallets
3. Generates a unique Flow wallet for each user (multi-threaded)
4. Saves private keys to flow/accounts/ directory
5. Saves wallet data to the database

Usage:
    python3 generate_flow_accounts.py
"""

import subprocess
import json
import os
import sys
import uuid
import time
import threading
import queue
import signal
from pathlib import Path
from datetime import datetime
from supabase import create_client, Client
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

# Configuration
NETWORK = "mainnet"
TRANSACTION_TIMEOUT = 300  # 5 minutes timeout for transaction sealing
RATE_LIMIT_DELAY = 1.0  # Delay between requests to avoid rate limiting
MAX_WORKERS = 1  # Number of worker threads for wallet generation (must be 1 to avoid sequence conflicts)

# Supabase configuration
SUPABASE_URL = os.getenv('SUPABASE_URL')
SUPABASE_SERVICE_KEY = os.getenv('SUPABASE_SERVICE_ROLE_KEY')

class FlowWalletGenerator:
    def __init__(self):
        self.running = True
        self.supabase = None
        self.flow_dir = Path("flow")
        self.flow_binary = None
        self.lock = threading.Lock()  # Thread safety lock
        self.transaction_queue = queue.Queue()  # Queue for sequential transaction processing
        self.transaction_worker = None  # Single worker thread for transactions
        self.wallets_data = {}  # Store generated wallet data
        self.successful_wallets = 0
        self.database_saves = 0
        self.skipped_wallets = 0
        
        # Setup signal handlers for graceful shutdown
        signal.signal(signal.SIGINT, self.signal_handler)
        signal.signal(signal.SIGTERM, self.signal_handler)
    
    def signal_handler(self, signum, frame):
        """Handle shutdown signals gracefully"""
        print(f"\n🛑 Received signal {signum}, shutting down gracefully...")
        self.running = False
        # Signal the transaction worker to stop
        if self.transaction_worker and self.transaction_worker.is_alive():
            self.transaction_queue.put(None)  # Sentinel value to stop worker
    
    def get_supabase_client(self):
        """Initialize and return Supabase client with service role"""
        try:
            if not SUPABASE_URL:
                print("Error: SUPABASE_URL not set in .env file")
                return None
                
            if not SUPABASE_SERVICE_KEY:
                print("Error: SUPABASE_SERVICE_ROLE_KEY not set in .env file")
                return None
            
            supabase: Client = create_client(SUPABASE_URL, SUPABASE_SERVICE_KEY)
            return supabase
        except Exception as e:
            print(f"Error initializing Supabase client: {e}")
            return None
    
    def get_flow_binary(self):
        """Get the Flow CLI binary path"""
        if self.flow_binary:
            return self.flow_binary
            
        try:
            result = subprocess.run(['which', 'flow'], capture_output=True, text=True)
            if result.returncode == 0:
                self.flow_binary = result.stdout.strip()
                return self.flow_binary
            else:
                print("Error: Flow CLI not found. Please install Flow CLI.")
                return None
        except Exception as e:
            print(f"Error finding Flow CLI: {e}")
            return None

    def get_existing_wallets(self):
        """Get all existing wallet auth_ids from the database"""
        try:
            existing_wallets = set()
            page = 1
            per_page = 1000
            
            while True:
                result = self.supabase.table('wallet').select('auth_id').range(
                    (page - 1) * per_page, 
                    page * per_page - 1
                ).execute()
                
                if not result.data or len(result.data) == 0:
                    break
                
                for wallet in result.data:
                    existing_wallets.add(wallet['auth_id'])
                
                if len(result.data) < per_page:
                    break
                    
                page += 1
            
            print(f"Found {len(existing_wallets)} existing wallets in database")
            return existing_wallets
            
        except Exception as e:
            print(f"Error fetching existing wallets from database: {e}")
            return set()
    
    def get_all_users(self):
        """Get all users from Supabase auth with pagination, excluding those with existing wallets"""
        try:
            # First get existing wallets
            existing_wallets = self.get_existing_wallets()
            
            all_users = []
            page = 1
            per_page = 1000
            total_users_fetched = 0
            
            print(f"🔍 Starting to fetch users from Supabase auth...")
            print(f"📊 Found {len(existing_wallets)} existing wallets to skip")
            
            while True:
                print(f"📄 Fetching page {page} (per_page={per_page})...")
                result = self.supabase.auth.admin.list_users(per_page=per_page, page=page)
                
                if not result or len(result) == 0:
                    print(f"📄 Page {page} returned no users, stopping pagination")
                    break
                
                print(f"📄 Page {page} returned {len(result)} users")
                total_users_fetched += len(result)
                
                users_batch = []
                for user in result:
                    # Skip users who already have wallets
                    if user.id in existing_wallets:
                        self.skipped_wallets += 1
                        continue
                    
                    users_batch.append({
                        'id': user.id,
                        'auth_id': user.id,
                        'created_at': user.created_at
                    })
                
                print(f"📄 Page {page}: {len(users_batch)} users need wallets (skipped {len(result) - len(users_batch)} with existing wallets)")
                all_users.extend(users_batch)
                
                if len(result) < per_page:
                    print(f"📄 Page {page} had fewer users than per_page, stopping pagination")
                    break
                    
                page += 1
            
            print(f"📊 Total users fetched from auth: {total_users_fetched}")
            print(f"📊 Users needing wallets: {len(all_users)}")
            print(f"📊 Users skipped (already have wallets): {self.skipped_wallets}")
            return all_users
            
        except Exception as e:
            print(f"Error fetching users from Supabase auth: {e}")
            return []

    def generate_flow_wallet(self):
        """Generate a single Flow wallet using Flow CLI"""
        try:
            flow_binary = self.get_flow_binary()
            if not flow_binary:
                return None
            
            # First, generate a key pair
            cmd = f"{flow_binary} keys generate -o json"
            result = subprocess.run(
                cmd,
                cwd=self.flow_dir,
                capture_output=True,
                text=True,
                shell=True,
                timeout=60
            )
            
            if result.returncode != 0:
                print(f"Failed to generate key pair: {result.stderr}")
                return None
            
            key_data = json.loads(result.stdout.strip())
            public_key = key_data['public']
            
            # Now create an actual Flow account with the public key
            cmd = f"{flow_binary} accounts create --key {public_key} --network {NETWORK} -o json"
            result = subprocess.run(
                cmd,
                cwd=self.flow_dir,
                capture_output=True,
                text=True,
                shell=True,
                timeout=60
            )
            
            if result.returncode != 0:
                print(f"Failed to create account: {result.stderr}")
                return None
            
            account_data = json.loads(result.stdout.strip())
            account_address = account_data['address']
            
            # Return the key data with the actual account address
            return {
                'private': key_data['private'],
                'public': key_data['public'],
                'address': account_address
            }
            
        except Exception as e:
            print(f"Error generating wallet: {e}")
            return None
    
    def wait_for_transaction_seal(self, tx_id):
        """Wait for a transaction to be sealed"""
        try:
            flow_binary = self.get_flow_binary()
            if not flow_binary:
                return False
            
            start_time = time.time()
            while time.time() - start_time < TRANSACTION_TIMEOUT:
                cmd = f"{flow_binary} transactions get {tx_id} --network {NETWORK} -o json"
                result = subprocess.run(
                    cmd,
                    cwd=self.flow_dir,
                    capture_output=True,
                    text=True,
                    shell=True,
                    timeout=30
                )
                
                if result.returncode == 0:
                    try:
                        tx_data = json.loads(result.stdout.strip())
                        status = tx_data.get("status", "")
                        if status == "SEALED":
                            return True
                        elif status == "FAILED":
                            print(f"Transaction {tx_id} failed")
                            return False
                    except (ValueError, KeyError):
                        pass
                
                time.sleep(5)  # Wait 5 seconds before checking again
            
            print(f"Transaction {tx_id} timed out after {TRANSACTION_TIMEOUT} seconds")
            return False
            
        except Exception as e:
            print(f"Error waiting for transaction seal: {e}")
            return False
    
    def transaction_worker_thread(self):
        """Single worker thread that processes transactions sequentially"""
        while self.running:
            try:
                # Get next transaction from queue (blocks until available)
                task = self.transaction_queue.get(timeout=1)
                
                # Check for shutdown signal
                if task is None:
                    break
                
                user, key_data = task
                auth_id = user['auth_id']
                
                with self.lock:
                    print(f"💾 Processing database save for {auth_id}")
                
                # Save to database
                if self.save_wallet_to_database(auth_id, key_data):
                    with self.lock:
                        self.database_saves += 1
                        self.wallets_data[auth_id] = key_data
                        self.successful_wallets += 1
                        print(f"✅ Successfully saved wallet for {auth_id}: {key_data['address']}")
                else:
                    with self.lock:
                        print(f"❌ Failed to save wallet to database for {auth_id}")
                
                # Mark task as done
                self.transaction_queue.task_done()
                
            except queue.Empty:
                # Timeout waiting for task, continue loop
                continue
            except Exception as e:
                with self.lock:
                    print(f"❌ Error in transaction worker: {e}")
                # Mark task as done even if failed
                try:
                    self.transaction_queue.task_done()
                except ValueError:
                    pass  # Task was already marked as done

    def save_private_key(self, auth_id, private_key):
        """Save private key to .pkey file"""
        accounts_dir = self.flow_dir / "accounts"
        accounts_dir.mkdir(exist_ok=True)
        
        pkey_file = accounts_dir / f"{auth_id}.pkey"
        with open(pkey_file, 'w') as f:
            f.write(private_key)
        
        print(f"✓ Saved private key: {pkey_file}")

    def save_wallet_to_database(self, auth_id, key_data):
        """Save wallet data to the database using recommended pattern"""
        try:
            wallet_data = {
                'id': str(uuid.uuid4()),
                'created_at': datetime.now().isoformat(),
                'auth_id': auth_id,
                'flow_address': key_data['address'],
                'flow_private_key': key_data['private'],
                'flow_public_key': key_data['public']
            }
            
            result = self.supabase.table('wallet').insert(wallet_data).execute()
            
            if result.data:
                return True
            else:
                return False
                
        except Exception as e:
            print(f"Error saving wallet to database for user {auth_id}: {e}")
            return False

    def create_flow_production_config(self):
        """Create flow-production.json with all generated wallets"""
        accounts_dir = self.flow_dir / "accounts"
        accounts_dir.mkdir(exist_ok=True)
        
        production_config = {
            "accounts": {}
        }
        
        for auth_id, wallet_data in self.wallets_data.items():
            production_config["accounts"][auth_id] = {
                "address": wallet_data['address'],
                "key": {
                    "type": "file",
                    "location": f"{auth_id}.pkey",
                    "signatureAlgorithm": "ECDSA_secp256k1",
                    "hashAlgorithm": "SHA2_256"
                }
            }
        
        # Write flow-production.json
        production_file = accounts_dir / "flow-production.json"
        with open(production_file, 'w') as f:
            json.dump(production_config, f, indent=4)
        
        print(f"✓ Created flow-production.json with {len(production_config['accounts'])} accounts")
        return production_file
    
    def process_user_wallet(self, user):
        """Process a single user - generate wallet and queue for database save"""
        auth_id = user['auth_id']
        
        # Add rate limiting delay
        time.sleep(RATE_LIMIT_DELAY)
        
        print(f"🔑 Generating wallet for {auth_id}")
        
        # Generate Flow wallet (sequential processing avoids sequence conflicts)
        key_data = self.generate_flow_wallet()
        
        if key_data:
            print(f"✅ Generated wallet for {auth_id}: {key_data['address']}")
            
            # Save private key immediately
            self.save_private_key(auth_id, key_data['private'])
            
            # Queue the database save for sequential processing
            self.transaction_queue.put((user, key_data))
            
            return True
        else:
            print(f"❌ Failed to generate wallet for {auth_id}")
            return False
    
    def run(self):
        """Main wallet generation process with sequential processing"""
        print("🚀 Starting Flow wallet generation for all users...")
        print("Wallet generation will be processed sequentially to avoid Flow sequence number conflicts")
        print("Database saves will be processed sequentially to avoid conflicts")
        
        # Ensure we're in the right directory
        if not self.flow_dir.exists():
            print("Error: flow directory not found. Please run this script from the project root.")
            sys.exit(1)
        
        # Initialize Supabase client
        self.supabase = self.get_supabase_client()
        if not self.supabase:
            print("Error: Could not initialize Supabase client.")
            print("Set SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY environment variables.")
            sys.exit(1)
        
        # Check if Flow CLI is available
        if not self.get_flow_binary():
            print("Error: Flow CLI not found. Please install Flow CLI.")
            sys.exit(1)
        
        # Start the transaction worker thread
        self.transaction_worker = threading.Thread(target=self.transaction_worker_thread, daemon=True)
        self.transaction_worker.start()
        
        # Get all users (excluding those with existing wallets)
        users = self.get_all_users()
        if not users:
            print("No users found that need wallets.")
            sys.exit(0)
        
        print(f"📊 Processing {len(users)} users...")
        
        # Process users sequentially to avoid Flow sequence number conflicts
        for user in users:
            if not self.running:
                break
                
            try:
                self.process_user_wallet(user)
            except Exception as e:
                print(f"❌ Error processing user {user['auth_id']}: {e}")
        
        # Wait for all queued database saves to complete
        if self.successful_wallets > 0:
            print(f"⏳ Waiting for {self.successful_wallets} database saves to complete...")
            self.transaction_queue.join()  # Wait for all tasks to be processed
        
        print(f"\n🎉 Summary:")
        print(f"- Processed: {len(users)} users")
        print(f"- Skipped: {self.skipped_wallets} users (already had wallets)")
        print(f"- Generated: {self.successful_wallets} wallets")
        print(f"- Saved to database: {self.database_saves} wallets")
        
        if self.successful_wallets > 0:
            # Create flow-production.json
            production_file = self.create_flow_production_config()
            print(f"- Configuration saved to {production_file}")
            
            # Show first few wallets as example
            print(f"\nFirst 3 wallets:")
            for i, (auth_id, data) in enumerate(list(self.wallets_data.items())[:3]):
                print(f"  {auth_id}: {data['address']}")
        else:
            print("No wallets were generated successfully.")
            sys.exit(1)

def main():
    generator = FlowWalletGenerator()
    generator.run()

if __name__ == "__main__":
    main()

File: generateFlowAccounts.js
#!/usr/bin/env node
/**
 * Flow Wallet Generator for All Users
 *
 * This script generates Flow wallets for all users in Supabase auth:
 * 1. Fetches all users from Supabase auth using service role
 * 2. Skips users who already have wallets
 * 3. Generates a unique Flow wallet for each user (multi-threaded)
 * 4. Saves private keys to flow/accounts/ directory
 * 5. Saves wallet data to the database
 *
 * Usage:
 *    node generateFlowAccounts.js
 */

const { createClient } = require('@supabase/supabase-js');
const { config } = require('dotenv');
const { exec } = require('child_process');
const { promisify } = require('util');
const fs = require('fs').promises;
const path = require('path');
const { v4: uuidv4 } = require('uuid');

const execAsync = promisify(exec);

// Load environment variables from .env file
config();

// Configuration
const NETWORK = "mainnet";
const TRANSACTION_TIMEOUT = 300; // 5 minutes timeout for transaction sealing
const RATE_LIMIT_DELAY = 1.0; // Delay between requests to avoid rate limiting
const MAX_WORKERS = 1; // Number of worker threads for wallet generation (must be 1 to avoid sequence conflicts)

// Supabase configuration
const SUPABASE_URL = process.env.SUPABASE_URL;
const SUPABASE_SERVICE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY;

class FlowWalletGenerator {
    constructor() {
        this.running = true;
        this.supabase = null;
        this.flowDir = './flow';
        this.flowBinary = null;
        this.transactionQueue = [];
        this.processingTransaction = false;
        this.walletsData = {};
        this.successfulWallets = 0;
        this.databaseSaves = 0;
        this.skippedWallets = 0;
        
        // Setup signal handlers for graceful shutdown
        process.on('SIGINT', () => this.signalHandler('SIGINT'));
        process.on('SIGTERM', () => this.signalHandler('SIGTERM'));
    }

    signalHandler(signal) {
        console.log(`\n🛑 Received signal ${signal}, shutting down gracefully...`);
        this.running = false;
    }

    async getSupabaseClient() {
        try {
            if (!SUPABASE_URL) {
                console.log("Error: SUPABASE_URL not set in .env file");
                return null;
            }
                
            if (!SUPABASE_SERVICE_KEY) {
                console.log("Error: SUPABASE_SERVICE_ROLE_KEY not set in .env file");
                return null;
            }
            
            const supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_KEY);
            return supabase;
        } catch (error) {
            console.log(`Error initializing Supabase client: ${error.message}`);
            return null;
        }
    }

    async getFlowBinary() {
        if (this.flowBinary) {
            return this.flowBinary;
        }
            
        try {
            const { stdout } = await execAsync('which flow');
            this.flowBinary = stdout.trim();
            return this.flowBinary;
        } catch (error) {
            console.log("Error: Flow CLI not found. Please install Flow CLI.");
            return null;
        }
    }

    async getExistingWallets() {
        try {
            const existingWallets = new Set();
            let page = 1;
            const perPage = 1000;
            
            while (true) {
                const { data, error } = await this.supabase
                    .from('wallet')
                    .select('auth_id')
                    .range((page - 1) * perPage, page * perPage - 1);
                
                if (error) {
                    console.log(`Error fetching existing wallets: ${error.message}`);
                    break;
                }
                
                if (!data || data.length === 0) {
                    break;
                }
                
                for (const wallet of data) {
                    existingWallets.add(wallet.auth_id);
                }
                
                if (data.length < perPage) {
                    break;
                }
                    
                page += 1;
            }
            
            console.log(`Found ${existingWallets.size} existing wallets in database`);
            return existingWallets;
            
        } catch (error) {
            console.log(`Error fetching existing wallets from database: ${error.message}`);
            return new Set();
        }
    }

    async getAllUsers() {
        try {
            // First get existing wallets
            const existingWallets = await this.getExistingWallets();
            
            const allUsers = [];
            let page = 1;
            const perPage = 1000;
            let totalUsersFetched = 0;
            
            console.log(`🔍 Starting to fetch users from Supabase auth...`);
            console.log(`📊 Found ${existingWallets.size} existing wallets to skip`);
            
            while (true) {
                console.log(`📄 Fetching page ${page} (per_page=${perPage})...`);
                
                const { data: users, error } = await this.supabase.auth.admin.listUsers({
                    page: page,
                    perPage: perPage
                });
                
                if (error) {
                    console.log(`Error fetching users: ${error.message}`);
                    break;
                }
                
                if (!users || users.length === 0) {
                    console.log(`📄 Page ${page} returned no users, stopping pagination`);
                    break;
                }
                
                console.log(`📄 Page ${page} returned ${users.length} users`);
                totalUsersFetched += users.length;
                
                const usersBatch = [];
                for (const user of users) {
                    // Skip users who already have wallets
                    if (existingWallets.has(user.id)) {
                        this.skippedWallets += 1;
                        continue;
                    }
                    
                    usersBatch.push({
                        id: user.id,
                        auth_id: user.id,
                        created_at: user.created_at
                    });
                }
                
                console.log(`📄 Page ${page}: ${usersBatch.length} users need wallets (skipped ${users.length - usersBatch.length} with existing wallets)`);
                allUsers.push(...usersBatch);
                
                if (users.length < perPage) {
                    console.log(`📄 Page ${page} had fewer users than per_page, stopping pagination`);
                    break;
                }
                    
                page += 1;
            }
            
            console.log(`📊 Total users fetched from auth: ${totalUsersFetched}`);
            console.log(`📊 Users needing wallets: ${allUsers.length}`);
            console.log(`📊 Users skipped (already have wallets): ${this.skippedWallets}`);
            return allUsers;
            
        } catch (error) {
            console.log(`Error fetching users from Supabase auth: ${error.message}`);
            return [];
        }
    }

    async generateFlowWallet() {
        try {
            const flowBinary = await this.getFlowBinary();
            if (!flowBinary) {
                return null;
            }
            
            // First, generate a key pair
            const keyCmd = `${flowBinary} keys generate -o json`;
            const { stdout: keyOutput } = await execAsync(keyCmd, {
                cwd: this.flowDir,
                timeout: 60000
            });
            
            const keyData = JSON.parse(keyOutput.trim());
            const publicKey = keyData.public;
            
            // Now create an actual Flow account with the public key
            const accountCmd = `${flowBinary} accounts create --key ${publicKey} --network ${NETWORK} -o json`;
            const { stdout: accountOutput } = await execAsync(accountCmd, {
                cwd: this.flowDir,
                timeout: 60000
            });
            
            const accountData = JSON.parse(accountOutput.trim());
            const accountAddress = accountData.address;
            
            // Return the key data with the actual account address
            return {
                private: keyData.private,
                public: keyData.public,
                address: accountAddress
            };
            
        } catch (error) {
            console.log(`Error generating wallet: ${error.message}`);
            return null;
        }
    }

    async waitForTransactionSeal(txId) {
        try {
            const flowBinary = await this.getFlowBinary();
            if (!flowBinary) {
                return false;
            }
            
            const startTime = Date.now();
            while (Date.now() - startTime < TRANSACTION_TIMEOUT * 1000) {
                const cmd = `${flowBinary} transactions get ${txId} --network ${NETWORK} -o json`;
                const { stdout } = await execAsync(cmd, {
                    cwd: this.flowDir,
                    timeout: 30000
                });
                
                try {
                    const txData = JSON.parse(stdout.trim());
                    const status = txData.status;
                    if (status === "SEALED") {
                        return true;
                    } else if (status === "FAILED") {
                        console.log(`Transaction ${txId} failed`);
                        return false;
                    }
                } catch (parseError) {
                    // Continue waiting
                }
                
                await new Promise(resolve => setTimeout(resolve, 5000)); // Wait 5 seconds before checking again
            }
            
            console.log(`Transaction ${txId} timed out after ${TRANSACTION_TIMEOUT} seconds`);
            return false;
            
        } catch (error) {
            console.log(`Error waiting for transaction seal: ${error.message}`);
            return false;
        }
    }

    async processTransactionQueue() {
        if (this.processingTransaction || this.transactionQueue.length === 0) {
            return;
        }

        this.processingTransaction = true;
        
        while (this.transactionQueue.length > 0 && this.running) {
            const task = this.transactionQueue.shift();
            const { user, keyData } = task;
            const authId = user.auth_id;
            
            console.log(`💾 Processing database save for ${authId}`);
            
            // Save to database
            if (await this.saveWalletToDatabase(authId, keyData)) {
                this.databaseSaves += 1;
                this.walletsData[authId] = keyData;
                this.successfulWallets += 1;
                console.log(`✅ Successfully saved wallet for ${authId}: ${keyData.address}`);
            } else {
                console.log(`❌ Failed to save wallet to database for ${authId}`);
            }
        }
        
        this.processingTransaction = false;
    }

    async savePrivateKey(authId, privateKey) {
        const accountsDir = path.join(this.flowDir, 'accounts');
        await fs.mkdir(accountsDir, { recursive: true });
        
        const pkeyFile = path.join(accountsDir, `${authId}.pkey`);
        await fs.writeFile(pkeyFile, privateKey);
        
        console.log(`✓ Saved private key: ${pkeyFile}`);
    }

    async saveWalletToDatabase(authId, keyData) {
        try {
            const walletData = {
                id: uuidv4(),
                created_at: new Date().toISOString(),
                auth_id: authId,
                flow_address: keyData.address,
                flow_private_key: keyData.private,
                flow_public_key: keyData.public
            };
            
            const { data, error } = await this.supabase
                .from('wallet')
                .insert(walletData);
            
            if (error) {
                console.log(`Error saving wallet to database for user ${authId}: ${error.message}`);
                return false;
            }
            
            return !!data;
                
        } catch (error) {
            console.log(`Error saving wallet to database for user ${authId}: ${error.message}`);
            return false;
        }
    }

    async createFlowProductionConfig() {
        const accountsDir = path.join(this.flowDir, 'accounts');
        await fs.mkdir(accountsDir, { recursive: true });
        
        const productionConfig = {
            accounts: {}
        };
        
        for (const [authId, walletData] of Object.entries(this.walletsData)) {
            productionConfig.accounts[authId] = {
                address: walletData.address,
                key: {
                    type: "file",
                    location: `${authId}.pkey`,
                    signatureAlgorithm: "ECDSA_secp256k1",
                    hashAlgorithm: "SHA2_256"
                }
            };
        }
        
        // Write flow-production.json
        const productionFile = path.join(accountsDir, 'flow-production.json');
        await fs.writeFile(productionFile, JSON.stringify(productionConfig, null, 4));
        
        console.log(`✓ Created flow-production.json with ${Object.keys(productionConfig.accounts).length} accounts`);
        return productionFile;
    }

    async processUserWallet(user) {
        const authId = user.auth_id;
        
        // Add rate limiting delay
        await new Promise(resolve => setTimeout(resolve, RATE_LIMIT_DELAY * 1000));
        
        console.log(`🔑 Generating wallet for ${authId}`);
        
        // Generate Flow wallet (sequential processing avoids sequence conflicts)
        const keyData = await this.generateFlowWallet();
        
        if (keyData) {
            console.log(`✅ Generated wallet for ${authId}: ${keyData.address}`);
            
            // Save private key immediately
            await this.savePrivateKey(authId, keyData.private);
            
            // Queue the database save for sequential processing
            this.transactionQueue.push({ user, keyData });
            
            return true;
        } else {
            console.log(`❌ Failed to generate wallet for ${authId}`);
            return false;
        }
    }

    async run() {
        console.log("🚀 Starting Flow wallet generation for all users...");
        console.log("Wallet generation will be processed sequentially to avoid Flow sequence number conflicts");
        console.log("Database saves will be processed sequentially to avoid conflicts");
        
        // Ensure we're in the right directory
        const fs = require('fs');
        if (!fs.existsSync(this.flowDir)) {
            console.log("Error: flow directory not found. Please run this script from the project root.");
            process.exit(1);
        }
        
        // Initialize Supabase client
        this.supabase = await this.getSupabaseClient();
        if (!this.supabase) {
            console.log("Error: Could not initialize Supabase client.");
            console.log("Set SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY environment variables.");
            process.exit(1);
        }
        
        // Check if Flow CLI is available
        if (!(await this.getFlowBinary())) {
            console.log("Error: Flow CLI not found. Please install Flow CLI.");
            process.exit(1);
        }
        
        // Get all users (excluding those with existing wallets)
        const users = await this.getAllUsers();
        if (!users || users.length === 0) {
            console.log("No users found that need wallets.");
            process.exit(0);
        }
        
        console.log(`📊 Processing ${users.length} users...`);
        
        // Process users sequentially to avoid Flow sequence number conflicts
        for (const user of users) {
            if (!this.running) {
                break;
            }
                
            try {
                await this.processUserWallet(user);
            } catch (error) {
                console.log(`❌ Error processing user ${user.auth_id}: ${error.message}`);
            }
        }
        
        // Wait for all queued database saves to complete
        if (this.successfulWallets > 0) {
            console.log(`⏳ Waiting for ${this.successfulWallets} database saves to complete...`);
            await this.processTransactionQueue();
        }
        
        console.log(`\n🎉 Summary:`);
        console.log(`- Processed: ${users.length} users`);
        console.log(`- Skipped: ${this.skippedWallets} users (already had wallets)`);
        console.log(`- Generated: ${this.successfulWallets} wallets`);
        console.log(`- Saved to database: ${this.databaseSaves} wallets`);
        
        if (this.successfulWallets > 0) {
            // Create flow-production.json
            const productionFile = await this.createFlowProductionConfig();
            console.log(`- Configuration saved to ${productionFile}`);
            
            // Show first few wallets as example
            console.log(`\nFirst 3 wallets:`);
            const walletEntries = Object.entries(this.walletsData);
            for (let i = 0; i < Math.min(3, walletEntries.length); i++) {
                const [authId, data] = walletEntries[i];
                console.log(`  ${authId}: ${data.address}`);
            }
        } else {
            console.log("No wallets were generated successfully.");
            process.exit(1);
        }
    }
}

async function main() {
    const generator = new FlowWalletGenerator();
    await generator.run();
}

if (require.main === module) {
    main().catch(console.error);
}

module.exports = { FlowWalletGenerator };
